File: src/redis/coordination.lua
local operation = ARGV[1]
local maximum = 9007199254740991
local batchSize = 500

local function member(page, id)
    return string.format('%016.0f:%016.0f', page, id)
end

local function problem(message)
    return { 'error', message }
end

local function operationId(value)
    return type(value) == 'string' and #value == 32 and string.match(value, '^[a-f0-9]+$') ~= nil
end

local function indexName(value)
    return type(value) == 'string' and string.match(value, '^[A-Za-z0-9_-]+$') ~= nil
end

local function integer(value)
    return type(value) == 'number' and value >= 0 and value <= maximum and value == math.floor(value)
end

local function validState(state)
    if type(state) ~= 'table' or state.version ~= 2 or not operationId(state.epoch)
        or type(state.closing) ~= 'boolean' or state.run == nil or state.pending == nil then return false end
    local run = state.run
    if run ~= cjson.null then
        if type(run) ~= 'table' or not operationId(run.id) or type(run.recovery) ~= 'boolean'
            or state.closing or not integer(run.watermark) or not indexName(run.completion)
            or (run.full ~= cjson.null and not indexName(run.full)) or run.full == run.completion then return false end
        local phases = { building = true, ready = true, catchup = true, swapping = true,
            activated = true, aborting = true }
        if type(run.phase) ~= 'string' or not phases[run.phase] then return false end
    end
    local pending = state.pending
    if pending == cjson.null then return true end
    if type(pending) ~= 'table' or not operationId(pending.id)
        or (pending.purpose ~= 'write' and pending.purpose ~= 'guard-delete')
        or state.closing ~= (pending.purpose == 'guard-delete')
        or (pending.method ~= 'POST' and pending.method ~= 'PATCH' and pending.method ~= 'DELETE')
        or type(pending.path) ~= 'string' or string.sub(pending.path, 1, 1) ~= '/'
        or type(pending.startedAt) ~= 'string' or pending.startedAt == ''
        or (pending.taskUid ~= cjson.null and not integer(pending.taskUid))
        or pending.swaps == nil or pending.createdIndex == nil then return false end
    if pending.purpose == 'guard-delete' and (pending.method ~= 'DELETE'
        or not string.match(pending.path, '^/indexes/[^/?]+$')) then return false end
    if pending.path == '/swap-indexes' then
        if pending.method ~= 'POST' or type(pending.swaps) ~= 'table' or #pending.swaps == 0 then return false end
        for _, swap in ipairs(pending.swaps) do
            if type(swap) ~= 'table' or type(swap.indexes) ~= 'table' or #swap.indexes ~= 2
                or not indexName(swap.indexes[1]) or not indexName(swap.indexes[2])
                or swap.indexes[1] == swap.indexes[2] then return false end
        end
    elseif pending.swaps ~= cjson.null then return false end
    if pending.path == '/indexes' then
        if pending.method ~= 'POST' or not indexName(pending.createdIndex) then return false end
    elseif pending.createdIndex ~= cjson.null then return false end
    return true
end

local function journalEntry(key, raw)
    local valid, entry = pcall(cjson.decode, raw)
    if not valid or type(entry) ~= 'table' or type(entry.done) ~= 'boolean'
        or type(entry.title) ~= 'string' or type(entry.completionOnly) ~= 'boolean' then
        return nil
    end
    local pending = redis.call('ZSCORE', KEYS[5], key)
    local done = redis.call('ZSCORE', KEYS[6], key)
    if entry.done then
        if pending or not done or tonumber(done) ~= tonumber(string.sub(key, 18)) then return nil end
    elseif not pending or tonumber(pending) ~= 0 or done then
        return nil
    end
    return entry
end

if operation == 'lock' then
    local acquired = redis.call('SET', KEYS[1], ARGV[2], 'NX', 'PX', ARGV[3])
    return { 'ok', acquired and 1 or 0 }
end

if operation == 'unlock' then
    if redis.call('GET', KEYS[1]) == ARGV[2] then redis.call('DEL', KEYS[1]) end
    return { 'ok', 1 }
end

local guarded = operation == 'assert' or operation == 'initialize' or operation == 'discard'
    or operation == 'writeState' or operation == 'markDone' or operation == 'prune'
    or operation == 'seal' or operation == 'retire'
if guarded then
    if redis.call('GET', KEYS[1]) ~= ARGV[2] then return { 'lost' } end
    redis.call('PEXPIRE', KEYS[1], ARGV[3])
end
if operation == 'assert' then return { 'ok', 1 } end
if operation == 'discard' then
    redis.call('DEL', unpack(KEYS, 2))
    return { 'ok', 1 }
end

local present = 0
for i = 2, #KEYS do present = present + redis.call('EXISTS', KEYS[i]) end
if operation == 'initialize' then
    if present ~= 0 then return problem('Scope already contains coordination data; refusing to overwrite it.') end
    local decoded, initial = pcall(cjson.decode, ARGV[4])
    if not decoded or not validState(initial) or initial.closing or initial.pending ~= cjson.null then
        return problem('Invalid initial coordination state.')
    end
    redis.call('HSET', KEYS[2], 'state', ARGV[4], 'sequence', '0', 'entries', '0')
    return { 'ok', 1 }
end
if present == 0 then
    if operation == 'readState' or operation == 'append' then return { 'ok', 0 } end
    if operation == 'watermark' then return { 'ok', '0' } end
    if operation == 'pages' then return { 'ok', ARGV[4], 1, {} } end
    if operation == 'entries' then return { 'ok', {} } end
    return problem('No active coordination scope exists.')
end

local types = { 'string', 'hash', 'hash', 'zset', 'zset', 'zset' }
for i, key in ipairs(KEYS) do
    local actual = redis.call('TYPE', key).ok
    if actual ~= 'none' and actual ~= types[i] then return problem('Invalid coordination key type.') end
    if i > 1 and redis.call('PTTL', key) >= 0 then return problem('Active coordination data has an expiry.') end
end

local rawState = redis.call('HGET', KEYS[2], 'state')
local sequence = redis.call('HGET', KEYS[2], 'sequence')
local entryCount = redis.call('HGET', KEYS[2], 'entries')
if not rawState or not sequence or not entryCount then
    return problem('Active coordination state is incomplete; recover lost work with indexing senders stopped.')
end
local decoded, state = pcall(cjson.decode, rawState)
if not decoded or not validState(state) then
    return problem('Invalid coordination state; refusing writes.')
end
local sequenceNumber = tonumber(sequence)
if not string.match(sequence, '^%d+$') or not sequenceNumber or sequenceNumber < 0 or sequenceNumber > maximum
    or string.format('%.0f', sequenceNumber) ~= sequence then
    return problem('Invalid journal sequence; refusing writes.')
end
local count = redis.call('ZCARD', KEYS[4])
if string.format('%.0f', count) ~= entryCount or count > sequenceNumber or redis.call('HLEN', KEYS[3]) ~= count
    or redis.call('ZCARD', KEYS[5]) + redis.call('ZCARD', KEYS[6]) ~= count then
    return problem('Journal indexes are incomplete; refusing writes.')
end

if state.closing and count ~= 0 then return problem('Closing coordination scope contains journal work.') end

if operation == 'seal' then
    if state.run ~= cjson.null or state.pending ~= cjson.null or count ~= 0 then return { 'ok', 0 } end
    state.closing = true
    redis.call('HSET', KEYS[2], 'state', cjson.encode(state))
    return { 'ok', 1 }
end
if operation == 'retire' then
    if not state.closing or state.run ~= cjson.null or state.pending ~= cjson.null or count ~= 0 then
        return problem('Only a sealed, empty coordination scope can be retired.')
    end
    redis.call('DEL', unpack(KEYS, 2))
    return { 'ok', 1 }
end
if operation == 'readState' then return { 'ok', rawState } end
if operation == 'writeState' then
    local decodedNext, nextState = pcall(cjson.decode, ARGV[4])
    if not decodedNext or not validState(nextState) or nextState.epoch ~= state.epoch
        or nextState.closing ~= state.closing then return problem('Coordination epoch or closing state changed.') end
    redis.call('HSET', KEYS[2], 'state', ARGV[4])
    return { 'ok', 1 }
end
if operation == 'watermark' then return { 'ok', sequence } end

if operation == 'append' then
    if state.closing then return { 'ok', 0 } end
    if sequenceNumber >= maximum then return problem('Journal sequence exhausted; refusing to reuse IDs.') end
    local id = redis.call('HINCRBY', KEYS[2], 'sequence', 1)
    local entry = member(tonumber(ARGV[4]), id)
    redis.call('HSET', KEYS[3], entry, ARGV[5])
    redis.call('ZADD', KEYS[4], 0, entry)
    redis.call('ZADD', KEYS[5], 0, entry)
    redis.call('HINCRBY', KEYS[2], 'entries', 1)
    return { 'ok', string.format('%.0f', id) }
end

if operation == 'pages' then
    local after = tonumber(ARGV[4])
    local through = tonumber(ARGV[5])
    local index = ARGV[6] == '1' and KEYS[5] or KEYS[4]
    local limit = tonumber(ARGV[7])
    local pages = {}
    for _ = 1, batchSize do
        local first = redis.call('ZRANGEBYLEX', index, '(' .. member(after, maximum), '+', 'LIMIT', 0, 1)
        if #first == 0 then return { 'ok', string.format('%.0f', after), 1, pages } end
        after = tonumber(string.sub(first[1], 1, 16))
        if tonumber(string.sub(first[1], 18)) <= through then
            pages[#pages + 1] = string.format('%.0f', after)
            if #pages == limit then break end
        end
    end
    return { 'ok', string.format('%.0f', after), 0, pages }
end

if operation == 'entries' then
    local page = tonumber(ARGV[4])
    local through = tonumber(ARGV[5])
    local after = tonumber(ARGV[6])
    local found = redis.call('ZRANGEBYLEX', KEYS[4], '(' .. member(page, after),
        '[' .. member(page, through), 'LIMIT', 0, batchSize)
    local entries = {}
    for _, key in ipairs(found) do
        local value = redis.call('HGET', KEYS[3], key)
        if not value then return problem('Journal entry is missing; refusing writes.') end
        if not journalEntry(key, value) then return problem('Journal entry and indexes disagree; refusing writes.') end
        entries[#entries + 1] = { string.format('%.0f', tonumber(string.sub(key, 18))), value }
    end
    return { 'ok', entries }
end

if operation == 'markDone' then
    local changes = {}
    for i = 5, #ARGV do
        local id = tonumber(ARGV[i])
        local key = member(tonumber(ARGV[4]), id)
        local raw = redis.call('HGET', KEYS[3], key)
        if raw then
            local entry = journalEntry(key, raw)
            if not entry then return problem('Journal entry and indexes disagree; refusing writes.') end
            if not entry.done then
                entry.done = true
                changes[#changes + 1] = { key, id, cjson.encode(entry) }
            end
        end
    end
    for _, change in ipairs(changes) do
        redis.call('HSET', KEYS[3], change[1], change[3])
        redis.call('ZREM', KEYS[5], change[1])
        redis.call('ZADD', KEYS[6], change[2], change[1])
    end
    return { 'ok', 1 }
end

if operation == 'prune' then
    if state.run ~= cjson.null then return problem('A rebuild still needs completed journal history.') end
    local found = redis.call('ZRANGEBYSCORE', KEYS[6], '-inf', ARGV[4], 'LIMIT', 0, batchSize)
    if #found > 0 then
        redis.call('HDEL', KEYS[3], unpack(found))
        redis.call('ZREM', KEYS[4], unpack(found))
        redis.call('ZREM', KEYS[6], unpack(found))
        redis.call('HINCRBY', KEYS[2], 'entries', -#found)
    end
    return { 'ok', #found }
end

return problem('Unknown coordination operation.')

File: tests/integration/coordination.php
<?php

namespace FrauxSearch\Integration;

use FrauxSearch\RedisCoordinationStore;
use MediaWiki\Maintenance\Maintenance;
use RuntimeException;

$IP = getenv( 'MW_INSTALL_PATH' ) ?: dirname( __DIR__, 4 );
require_once "$IP/maintenance/Maintenance.php";

class CheckFrauxSearchCoordination extends Maintenance {
	public function __construct() {
		parent::__construct();
		$this->addDescription( 'Validate coordination with isolated Redis keys/indexes and real server faults.' );
		$this->addOption( 'execute', 'Create isolated test resources and deliberately expire only this harness writer lease' );
	}

	public function execute() {
		if ( !$this->hasOption( 'execute' ) ) {
			throw new RuntimeException( 'Pass --execute to run isolated Redis/Meilisearch mutation tests.' );
		}
		require_once __DIR__ . '/CoordinationRuntime.php';
		$runtime = new CoordinationRuntime();
		$this->output( json_encode( [ 'event' => 'coordination_integration_started',
			'index' => $runtime->index, 'scope_hash' => hash( 'sha256', $runtime->scope ),
			'redis_keys' => $runtime->store->keyNames(),
		], JSON_UNESCAPED_SLASHES ) . "\n" );
		try {
			$runtime->install();
			$probeRun = $runtime->coordinator()->beginRebuild( false );
			$this->checkLockLoss( $runtime );
			$this->output( "PASS separate-connection lease contention, forced expiry, and stale Lua owner rejection\n" );
			$this->checkAtomicJournal( $runtime );
			$this->output( "PASS atomic journal appends, independent producers, late append, and exact acknowledgements\n" );
			$runtime->coordinator()->abortRebuild( $probeRun['id'] );
			$this->checkLostWriteResponse( $runtime );
			$this->output( "PASS accepted HTTP write with discarded response, blocking, adoption, and replay\n" );
			$this->checkIdleCacheClear( $runtime );
			$this->output( "PASS idle scoped cache clear preserves completed indexes and permits a later refresh\n" );
			$this->checkActiveCacheLoss( $runtime );
			$this->output( "PASS active cache loss blocks writes and requires a complete recovered generation pair\n" );
			$this->checkLostGuardDeleteResponse( $runtime );
			$this->output( "PASS unknown guard DELETE adopts its exact real task without resubmission\n" );
			foreach ( [ false, true ] as $completionOnly ) {
				$this->checkCutover( $runtime, $completionOnly );
				$this->output( 'PASS ' . ( $completionOnly ? 'completion' : 'full' )
					. " generation replay, lost swap response, concurrent edit, and interrupted cleanup\n" );
			}
			$this->output( json_encode( [ 'event' => 'coordination_integration_passed',
				'accepted_tasks' => count( $runtime->http->accepted ), 'state' => $runtime->store->readState(),
			], JSON_UNESCAPED_SLASHES ) . "\n" );
		} finally {
			$runtime->cleanup();
			$this->output( json_encode( [ 'event' => 'coordination_integration_cleaned',
				'index' => $runtime->index, 'accepted_tasks_including_cleanup' => count( $runtime->http->accepted ),
				'task_uids' => array_column( $runtime->http->accepted, 'uid' ),
			], JSON_UNESCAPED_SLASHES ) . "\n" );
		}
		return true;
	}

	private function checkLockLoss( CoordinationRuntime $runtime ): void {
		$store = new RedisCoordinationStore( $runtime->connection->evaluate( ... ), $runtime->scope, 1000 );
		$successor = new RedisCoordinationStore( $runtime->observer->evaluate( ... ), $runtime->scope );
		$leaseKey = $store->keyNames()[0];
		CoordinationRuntime::check( $store->lock( 0 ), 'Initial writer cannot lock.' );
		try {
			$state = $store->readState();
			$store->writeState( $state + [ 'probe' => 'original' ] );
			CoordinationRuntime::check( !$successor->lock( 0 ), 'A separate connection acquired an already held lease.' );
			$ttl = $runtime->observer->evaluate( "return redis.call('PTTL', KEYS[1])", [ $leaseKey ], [] );
			CoordinationRuntime::check( is_int( $ttl ) && $ttl > 0 && $ttl <= 1000, 'Writer lease lacks the requested short TTL.' );
			$store->assertLocked();
			CoordinationRuntime::check( $runtime->observer->evaluate( "return redis.call('PEXPIRE', KEYS[1], 0)",
				[ $leaseKey ], [] ) === 1, 'Failed to expire this run\'s writer lease.' );
			CoordinationRuntime::check( $runtime->observer->evaluate( "return redis.call('PTTL', KEYS[1])",
				[ $leaseKey ], [] ) === -2, 'Expired writer lease remains present.' );
			CoordinationRuntime::check( $successor->lock( 0 ), 'Successor could not acquire the expired lease.' );
			$successor->writeState( $state + [ 'probe' => 'successor' ] );
			try {
				$store->writeState( $state + [ 'probe' => 'stale' ] );
				throw new RuntimeException( 'Stale writer unexpectedly retained its lease.' );
			} catch ( RuntimeException $error ) {
				CoordinationRuntime::check( str_contains( $error->getMessage(), 'lost its lease' ),
					'Expected lease-loss rejection, received: ' . $error->getMessage() );
			}
			CoordinationRuntime::check( $store->readState()['probe'] === 'successor',
				'Lua ownership condition allowed a stale writer to overwrite successor state.' );
			$store->unlock();
			$successor->assertLocked();
			CoordinationRuntime::check( !$runtime->store->lock( 0 ), 'Stale cleanup released the successor lease.' );
		} finally {
			$store->unlock();
			$successor->unlock();
		}
	}

	private function checkAtomicJournal( CoordinationRuntime $runtime ): void {
		$producer = new RedisCoordinationStore( $runtime->observer->evaluate( ... ), $runtime->scope );
		$firstId = $producer->append( 10, 'Observed completion title', true );
		CoordinationRuntime::check( is_int( $firstId ), 'Open probe run refused a journal append.' );
		CoordinationRuntime::check( $runtime->store->watermark() === $firstId,
			'Append sequence was not visible to the independent reader.' );
		$entries = $runtime->store->entries( 10, $firstId );
		CoordinationRuntime::check( array_keys( $entries ) === [ $firstId ]
			&& $entries[$firstId]['title'] === 'Observed completion title' && !$entries[$firstId]['done'],
			'Append did not publish its payload and journal indexes together.' );
		CoordinationRuntime::check( $runtime->store->lock( 0 ), 'Cannot acknowledge journal test.' );
		try {
			$lateId = $producer->append( 10, 'Late full-refresh title', false );
			$otherId = $runtime->store->append( 11, null, false );
			$through = $producer->watermark();
			CoordinationRuntime::check( $firstId < $lateId && $lateId < $otherId && $through === $otherId,
				'Independent producers did not allocate strictly increasing journal IDs.' );
			$runtime->store->markDone( 10, array_keys( $entries ) );
			$remaining = $producer->entries( 10, $through );
			CoordinationRuntime::check( array_keys( $remaining ) === [ $firstId, $lateId ] && $remaining[$firstId]['done']
				&& !$remaining[$lateId]['done']
				&& !$remaining[$lateId]['completionOnly'] && $remaining[$lateId]['title'] === 'Late full-refresh title',
				'Exact acknowledgement erased the independently appended full refresh.' );
			CoordinationRuntime::check( $producer->pages( 0, $through, true ) === [ 10, 11 ],
				'Acknowledgement erased another producer\'s pending page.' );
			$runtime->store->markDone( 10, [ $lateId ] );
			$runtime->store->markDone( 11, [ $otherId ] );
			CoordinationRuntime::check( $producer->pages( 0, $through, true ) === []
				&& $producer->pages( 0, $through, false ) === [ 10, 11 ],
				'Exact acknowledgements did not preserve the history needed by the probe run.' );
		} finally { $runtime->store->unlock(); }
	}

	private function checkLostWriteResponse( CoordinationRuntime $runtime ): void {
		$coordinator = $runtime->coordinator();
		$run = $coordinator->beginRebuild( false );
		$coordinator->ready( $run['id'] );
		$coordinator->finishRebuild( $run['id'] );
		$runtime->put( CoordinationRuntime::document( 1, 'before discarded acknowledgement' ) );
		$runtime->http->afterAccept = static function ( $method, $path ) use ( $runtime ): void {
			if ( $method === 'POST' && str_contains( $path, '/documents?' ) ) {
				$runtime->http->afterAccept = null;
				throw new InjectedFault( 'Discarded actual write acknowledgement.' );
			}
		};
		$this->expectInjected( fn () => $coordinator->refresh( 1 ) );
		$pending = $coordinator->status()['pending'];
		CoordinationRuntime::check( $pending !== null && $pending['taskUid'] === null, 'Lost response lacks durable unknown intent.' );
		$accepted = end( $runtime->http->accepted );
		$task = $runtime->client->getTask( $accepted['uid'] );
		CoordinationRuntime::check( $task['indexUid'] === $runtime->index, 'Observed real task belongs to a different index.' );
		$runtime->put( CoordinationRuntime::document( 1, 'new authoritative source state' ) );
		$count = count( $runtime->http->accepted );
		try {
			$coordinator->refresh( 1 );
			throw new RuntimeException( 'Unknown task did not block a successor.' );
		} catch ( RuntimeException $error ) {
			CoordinationRuntime::check( str_contains( $error->getMessage(), 'unknown outcome' ), 'Unexpected blocking failure.' );
		}
		CoordinationRuntime::check( count( $runtime->http->accepted ) === $count, 'Unknown mutation was resubmitted.' );
		$coordinator->adoptTask( $pending['id'], $accepted['uid'] );
		$coordinator->drain();
		$this->assertDocument( $runtime, 1, 'new authoritative source state' );
	}

	private function checkIdleCacheClear( CoordinationRuntime $runtime ): void {
		$this->assertIdle( $runtime );
		$cleared = $runtime->observer->evaluate( "return redis.call('DEL', unpack(KEYS))",
			$runtime->store->keyNames(), [] );
		CoordinationRuntime::check( $cleared === 0, 'Completed indexing retained Redis keys before the idle clear.' );
		$this->assertDocument( $runtime, 1, 'new authoritative source state' );
		$runtime->put( CoordinationRuntime::document( 1, 'refreshed after idle cache clear' ) );
		$runtime->coordinator()->refresh( 1 );
		$this->assertDocument( $runtime, 1, 'refreshed after idle cache clear' );
		$this->assertIdle( $runtime );
	}

	private function checkActiveCacheLoss( CoordinationRuntime $runtime ): void {
		$coordinator = $runtime->coordinator();
		$run = $coordinator->beginRebuild( false );
		$old = $runtime->build( 1 )['document'];
		$coordinator->writeBatch( $run['id'], [ $old ], [ $old ] );
		$runtime->put( CoordinationRuntime::document( 1, 'edited before active cache loss' ) );
		$runtime->put( CoordinationRuntime::document( 2, 'created before active cache loss' ) );
		$runtime->store->append( 1, null, false );
		$runtime->store->append( 2, null, false );
		$epoch = $runtime->store->readState()['epoch'];
		$removed = $runtime->observer->evaluate( "return redis.call('DEL', unpack(KEYS))",
			$runtime->store->keyNames(), [] );
		CoordinationRuntime::check( is_int( $removed ) && $removed > 0, 'Active cache-loss fault removed no state.' );
		$status = $coordinator->status();
		CoordinationRuntime::check( $status['coordinationLost'] && $status['guardEpoch'] === $epoch,
			'Lost run and journal did not leave a visible Meilisearch recovery guard.' );
		$count = count( $runtime->http->accepted );
		$this->expectFailure( fn () => $coordinator->refresh( 1 ), 'coordination was lost' );
		$this->expectFailure( fn () => $coordinator->writeBatch( $run['id'], [ $old ], [ $old ] ), 'coordination was lost' );
		CoordinationRuntime::check( count( $runtime->http->accepted ) === $count,
			'Lost active coordination allowed another HTTP mutation.' );
		$recovered = $coordinator->restartAfterStateLoss( $epoch );
		CoordinationRuntime::check( $recovered['recovery'] === true && $recovered['full'] !== null
			&& $runtime->store->readState()['epoch'] === $epoch,
			'Recovery did not atomically require a full rebuild under the original guard.' );
		$this->expectFailure( fn () => $coordinator->abortRebuild( $recovered['id'] ), 'cannot be aborted to idle' );
		foreach ( [ $run['full'], $run['completion'] ] as $index ) {
			CoordinationRuntime::check( !$runtime->client->withIndex( $index )->indexExists(),
				'An abandoned generation survived lost-state recovery.' );
		}
		$documents = [ $runtime->build( 1 )['document'], $runtime->build( 2 )['document'] ];
		$coordinator->writeBatch( $recovered['id'], $documents, $documents );
		$coordinator->ready( $recovered['id'] );
		$coordinator->finishRebuild( $recovered['id'] );
		$this->assertDocument( $runtime, 1, 'edited before active cache loss' );
		$this->assertDocument( $runtime, 2, 'created before active cache loss' );
		foreach ( [ $runtime->index, $runtime->index . '_completion' ] as $index ) {
			$page = $runtime->client->withIndex( $index )->listDocuments( 0, 100, [ 'id' ] );
			$ids = array_column( $page['results'], 'id' );
			sort( $ids, SORT_NUMERIC );
			CoordinationRuntime::check( $page['total'] === 2 && $ids === [ 1, 2 ],
				'Recovered generation did not contain exactly the complete source set.' );
		}
		$this->assertIdle( $runtime );
		$this->output( json_encode( [ 'event' => 'coordination_active_loss_recovered', 'epoch' => $epoch,
			'lost_run' => $run['id'], 'recovery_run' => $recovered['id'], 'removed_keys' => $removed,
		], JSON_UNESCAPED_SLASHES ) . "\n" );
	}

	private function checkLostGuardDeleteResponse( CoordinationRuntime $runtime ): void {
		$coordinator = $runtime->coordinator();
		$guardPath = '/indexes/' . rawurlencode( $runtime->index . '_coordination' );
		$runtime->put( CoordinationRuntime::document( 1, 'guard deletion acknowledgement lost' ) );
		$runtime->http->afterAccept = static function ( $method, $path ) use ( $runtime, $guardPath ): void {
			if ( $method === 'DELETE' && $path === $guardPath ) {
				$runtime->http->afterAccept = null;
				throw new InjectedFault( 'Discarded actual guard deletion acknowledgement.' );
			}
		};
		$this->expectInjected( fn () => $coordinator->refresh( 1 ) );
		$state = $runtime->store->readState();
		$pending = $state['pending'];
		CoordinationRuntime::check( $state['closing'] && $pending !== null
			&& $pending['purpose'] === 'guard-delete' && $pending['taskUid'] === null && $pending['path'] === $guardPath,
			'Unknown guard deletion did not retain a sealed scope and exact intent.' );
		$accepted = end( $runtime->http->accepted );
		$runtime->client->waitForTask( $accepted['uid'] );
		$task = $runtime->client->getTask( $accepted['uid'] );
		CoordinationRuntime::check( $task['type'] === 'indexDeletion'
			&& $task['indexUid'] === $runtime->index . '_coordination' && $task['status'] === 'succeeded',
			'Observed accepted task is not this epoch\'s completed guard deletion.' );
		$count = count( $runtime->http->accepted );
		$this->expectFailure( fn () => $coordinator->refresh( 1 ), 'unknown outcome' );
		CoordinationRuntime::check( count( $runtime->http->accepted ) === $count,
			'An unknown guard deletion was retried or a successor started early.' );
		$coordinator->adoptTask( $pending['id'], $accepted['uid'] );
		CoordinationRuntime::check( count( $runtime->http->accepted ) === $count,
			'Exact adoption resubmitted guard deletion instead of retiring the sealed scope.' );
		$this->assertDocument( $runtime, 1, 'guard deletion acknowledgement lost' );
		$this->assertIdle( $runtime );
		$this->output( json_encode( [ 'event' => 'coordination_guard_delete_adopted',
			'operation' => $pending['id'], 'task' => $accepted['uid'], 'resubmitted' => false,
		], JSON_UNESCAPED_SLASHES ) . "\n" );
	}

	private function checkCutover( CoordinationRuntime $runtime, bool $completionOnly ): void {
		$coordinator = $runtime->coordinator();
		$runtime->put( CoordinationRuntime::document( 1, 'generation old' ) );
		$runtime->put( CoordinationRuntime::document( 2, 'will be deleted' ) );
		$coordinator->refresh( 1 );
		$coordinator->refresh( 2 );
		$run = $coordinator->beginRebuild( $completionOnly );
		$documents = [ $runtime->build( 1 )['document'], $runtime->build( 2 )['document'] ];
		$coordinator->writeBatch( $run['id'], $documents, $documents );
		$runtime->put( CoordinationRuntime::document( 1, 'edited while generation built' ) );
		$runtime->remove( 2 );
		$runtime->put( CoordinationRuntime::document( 3, 'created while generation built' ) );
		foreach ( [ 1, 2, 3 ] as $id ) { $runtime->store->append( $id, null, false ); }
		$coordinator->ready( $run['id'] );
		$observerStore = new RedisCoordinationStore( $runtime->observer->evaluate( ... ), $runtime->scope );
		$runtime->http->afterAccept = static function ( $method, $path ) use ( $runtime, $observerStore ): void {
			if ( $path === '/swap-indexes' ) {
				$runtime->http->afterAccept = null;
				$runtime->put( CoordinationRuntime::document( 1, 'edited after accepted swap' ) );
				$observerStore->append( 1, 'Prior cutover title', false );
				throw new InjectedFault( 'Discarded actual swap acknowledgement after an independent Redis append.' );
			}
		};
		$beforeSwaps = count( array_filter( $runtime->http->accepted, static fn ( $task ) => $task['path'] === '/swap-indexes' ) );
		$this->expectInjected( fn () => $coordinator->finishRebuild( $run['id'] ) );
		$pending = $coordinator->status()['pending'];
		$swapTask = end( $runtime->http->accepted )['uid'];
		CoordinationRuntime::check( $pending['path'] === '/swap-indexes' && $pending['taskUid'] === null,
			'Accepted swap did not retain unknown outcome.' );
		$coordinator->adoptTask( $pending['id'], $swapTask );
		CoordinationRuntime::check( $coordinator->status()['run']['phase'] === 'activated', 'Recovered swap lacks durable activated phase.' );
		$runtime->http->afterAccept = static function ( $method, $path ) use ( $runtime, $run ): void {
			if ( $method === 'DELETE' && in_array( $path, [ '/indexes/' . $run['full'], '/indexes/' . $run['completion'] ], true ) ) {
				$runtime->http->afterAccept = null;
				throw new InjectedFault( 'Discarded actual generation cleanup acknowledgement.' );
			}
		};
		$this->expectInjected( fn () => $coordinator->finishRebuild( $run['id'] ) );
		$pending = $coordinator->status()['pending'];
		$deleteTask = end( $runtime->http->accepted )['uid'];
		$coordinator->adoptTask( $pending['id'], $deleteTask );
		$coordinator->finishRebuild( $run['id'] );
		$this->assertDocument( $runtime, 1, 'edited after accepted swap' );
		$this->assertDocument( $runtime, 3, 'created while generation built' );
		foreach ( [ $runtime->index, $runtime->index . '_completion' ] as $index ) {
			CoordinationRuntime::check( $runtime->client->withIndex( $index )->getDocument( 2 ) === null, 'Deleted source survived cutover.' );
		}
		foreach ( array_filter( [ $run['full'], $run['completion'] ] ) as $index ) {
			CoordinationRuntime::check( !$runtime->client->withIndex( $index )->indexExists(), 'Old generation survived cleanup recovery.' );
		}
		$swaps = array_filter( $runtime->http->accepted, static fn ( $task ) => $task['path'] === '/swap-indexes' );
		CoordinationRuntime::check( count( $swaps ) === $beforeSwaps + 1, 'Recovered cutover submitted another swap.' );
		CoordinationRuntime::check( $coordinator->status()['run'] === null && $coordinator->status()['pending'] === null
			&& $runtime->store->pages( 0, $runtime->store->watermark(), true ) === [], 'Cutover left a run, pending task, or pending journal.' );
	}

	private function assertDocument( CoordinationRuntime $runtime, int $id, string $text ): void {
		$expected = CoordinationRuntime::document( $id, $text )['document'];
		foreach ( [ $runtime->index, $runtime->index . '_completion' ] as $index ) {
			$actual = $runtime->client->withIndex( $index )->getDocument( $id );
			CoordinationRuntime::check( $actual !== null && $actual['document_hash'] === $expected['document_hash']
				&& $actual['revision_id'] === 1 && $actual['text'] === $text, 'Authoritative content/hash diverged after recovery.' );
		}
	}

	private function expectInjected( \Closure $callback ): void {
		try { $callback(); } catch ( InjectedFault ) { return; }
		throw new RuntimeException( 'Expected transport fault was never injected.' );
	}

	private function expectFailure( \Closure $callback, string $message ): void {
		try { $callback(); } catch ( RuntimeException $error ) {
			CoordinationRuntime::check( str_contains( $error->getMessage(), $message ),
				'Unexpected failure: ' . $error->getMessage() );
			return;
		}
		throw new RuntimeException( 'Expected rejection was not observed: ' . $message );
	}

	private function assertIdle( CoordinationRuntime $runtime ): void {
		CoordinationRuntime::check( $runtime->store->readState() === null
			&& !$runtime->client->withIndex( $runtime->index . '_coordination' )->indexExists()
			&& $runtime->observer->evaluate( "return redis.call('EXISTS', unpack(KEYS))",
				$runtime->store->keyNames(), [] ) === 0, 'Settled indexing retained its guard or Redis scope.' );
	}
}

$maintClass = CheckFrauxSearchCoordination::class;
require_once RUN_MAINTENANCE_IF_MAIN;
